cli/lib/kontena/cli/stacks/upgrade_command.rb

```diff
@@ -77,6 +82,66 @@ def merge_data(local_data, remote_data)
       merged
     end
 
+    def dry_report(merged)
+      added_services = []
+      removed_services = []
+      kept_services = {}
+      merged.each do |stackname, data|
+        if data[:remote].nil?
+          dry_msg "New stack #{stackname} would be installed to master"
+          added_services = data[:local][:stack]['services'].map { |svc| "#{stackname}/#{svc['name']}" }
+        elsif data[:local]
+          data[:local][:stack]['services'].each do |svc|
+            remote_svc = data[:remote]['services'].find { |s| s['name'] == svc['name'] }
+            if remote_svc
+              kept_services["#{stackname}/#{svc['name']}"] ||= {}
+              #kept_services["#{stackname}/#{svc['name']}"][:before] = JSON.load(JSON.dump(remote_svc)) # naughty deep clone/dup
+              #kept_services["#{stackname}/#{svc['name']}"][:after] = JSON.load(JSON.dump(svc))
+            else
+              added_services << "#{stackname}/#{svc['name']}"
+            end
+          end
+        end
+      end
+
+      merged.each do |stackname, data|
+        if data[:local].nil?
+          dry_msg "Stack  #{stackname} would be removed from master"
+          removed_services = data[:remote]['services'].map { |svc| "#{stackname}/#{svc['name']}" }
+        else
+          data[:local][:stack]['services'].each do |svc|
+            unless data[:remote]['services'].find { |s| s['name'] == svc['name'] }
```

Comment:
Broken for new stacks:

```
DRY RUN: New stack deptest-quux would be installed to master
bundler: failed to load command: bin/kontena (bin/kontena)
NoMethodError: undefined method `[]' for nil:NilClass
  /home/kontena/kontena/kontena/cli/lib/kontena/cli/stacks/upgrade_command.rb:113:in `block (2 levels) in dry_report'
  /home/kontena/kontena/kontena/cli/lib/kontena/cli/stacks/upgrade_command.rb:112:in `each'
  /home/kontena/kontena/kontena/cli/lib/kontena/cli/stacks/upgrade_command.rb:112:in `block in dry_report'
  /home/kontena/kontena/kontena/cli/lib/kontena/cli/stacks/upgrade_command.rb:107:in `each'
  /home/kontena/kontena/kontena/cli/lib/kontena/cli/stacks/upgrade_command.rb:107:in `dry_report'
  /home/kontena/kontena/kontena/cli/lib/kontena/cli/stacks/upgrade_command.rb:199:in `execute'
```

I suspect this block isn't even needed, because the `added_services` have already been collected earlier?